Find all conditionals of a required object

minhtule opened the pull request:
This module should be a library like `SendFile`, not a mixin. For example,
```ruby
achievement = Course::Achievement.first
conditionals_depending_on_the_achievement = Condition.conditionals_of(achievement)
```

lowjoel commented on app/models/course/condition.rb:
```diff
@@ -5,4 +5,12 @@ class Course::Condition < ActiveRecord::Base
+
+  def self.all_conditions
+    [
+      Course::Condition::Achievement,
```
I suggest not passing around class objects because in development autoloading will not have these objects see changes.

lowjoel replied:
Pass strings, then constantize.

lowjoel replied:
@minhtule is this needed now that there's the ALL_CONDITIONS constant?

lowjoel commented on app/models/course/condition/level.rb:
```diff
@@ -14,4 +14,8 @@ def title
   end
+
+  def self.required_classes
+    []
```
why is this empty?

minhtule replied:
It's because `Course::Condition::Level` doesn't depend on any class, unlike `Course::Condition::Achievement` depends on `Course::Achievement`.

lowjoel commented on app/models/course/condition/achievement.rb:
```diff
@@ -18,6 +18,10 @@ def satisfied_by?(course_user)
     course_user.achievements.exists?(achievement)
   end
 
+  def self.required_classes
```
Can you document the use?

lowjoel replied:
`dependent_classes` or something?

lowjoel commented on lib/autoload/condition.rb:
```diff
@@ -0,0 +1,35 @@
+    end.values
+  end
+
+  def self.required_class_to_condition_class_mapping
```
what... does this do?

lowjoel commented on lib/autoload/condition.rb:
```diff
@@ -0,0 +1,35 @@
+
+  private
+
+  def self.condition_classes_of(required_object)
```
How is this module mixed in?

minhtule commented on lib/autoload/condition.rb:
```diff
@@ -0,0 +1,44 @@
+  # @param [Course::Assessment, Course::Achievement] dependent_object An assessment or
+  #   achievement that conditionals depends on
+  # @return [Object] acts_as_conditional Objects that depend on the condition_object
+  def self.conditionals_of(dependent_object)
```
Fixed :grin:

minhtule commented:
@lowjoel I have rebased and added comments in to explain the functionalities of methods. They should address questions you had previously. Please take a look.

lowjoel commented:
@minhtule thanks! So if I understand what you're doing, you're trying to get a condition, then mapping the condition to the associated class?

How does this work for, say, assessments?

minhtule commented:
let say we have an Course::Assessment `assessment_1`, we would check the mapping and figure out that some Course::Condition::Assessment might depend on this `assessment_1`. Then we join the Course::Condition and Course::Condition::Assessment to find all conditionals that requires this `assessment_1` to be completed (with certain grade percentage).

lowjoel commented on app/models/course/condition/achievement.rb:
```diff
@@ -18,6 +18,11 @@ def satisfied_by?(course_user)
   end
 
+  # Array of classes that the condition depends on.
+  def self.dependent_classes
```
Should acts_as_condition define a dummy `dependent_classes` which raises NotImplementedError?

minhtule replied:
I couldn't figure out how to define a class method in `acts_as_condition`

lowjoel replied:
Class methods are methods in the class' singleton class.

So, to do that, define another module (`ClassMethods` by convention, so `ConditionClassMethods`) then `extend ClassMethods` after the `include ConditionInstanceMethods` line.

lowjoel commented on lib/autoload/condition.rb:
```diff
@@ -0,0 +1,43 @@
+    end.flatten.map(&:conditional)
+  end
+
+  private
```
This doesn't make the module methods private; private only applies to instance methods. If you want, you need to define the methods within the singleton class:

```ruby
class << self
  def conditionals_of
    # ..
  end

  private

  def condition_classes_of...
```

lowjoel commented on lib/autoload/condition.rb:
```diff
@@ -0,0 +1,43 @@
+  def self.dependent_class_to_condition_class_mapping
+    Course::Condition::ALL_CONDITIONS.map do |condition_name|
+      condition_name.constantize.dependent_classes.map do |dependent_name|
+        { dependent_name => condition_name }
```
What if the dependent class is depended on by different types of conditions? The hash will only keep one of the dependent classes.

eg Level condition depends on Level. LevelAndAchievement depends on Level.

lowjoel replied:
Or maybe I'm not too clear what's being done... this algorithm is rather complex.

minhtule replied:
You're right. Some how I kept imagining ( a => b ) is a pair >.< Will fix this part.

lowjoel commented:
Doesn't this just require querying the general condition table for all rows which are conditional of the object you have?
```ruby
Course::Condition.where(conditional: assessment_1)
```

minhtule commented:
No, we can't because the object is not a condition, it is an achievement or an assessment that some condition depends on.

lowjoel commented:
I see. So are you trying to answer the question "what will I unlock if I do this?"

minhtule commented:
Why should the logic be in the Course::Conditional? The object X might not be a conditional. Let say we have this Course::Condition::Forum that depends on a forum thread Course::Forum::Topic. Then the object X could be an instance of Course::Forum::Topic (posting on this thread will unlock some achievement). From object X, we need to figure out which types of condition could X be required by before finding the conditionals.

I can't think of any other usage right now. Any suggestion on the design? Otherwise, I'll fix the logic bug above :grin:

lowjoel commented:
I was thinking of Course::Conditional.find_conditions_for(obj), or Course::Condition.find_conditions_for(obj)?

I'm just not sure where else this will be used, if the only case is to show what attempting an object then either of the two places above make sense.

What do you think?

minhtule commented:
Ah, I see. The method is looking for conditionals and it needs to know the 2-way polymorphic relation of Course::Condition. So I think it should be Course::Condition.find_conditionals_for(obj). Are you okay with that?

lowjoel commented:
:+1: thanks!

minhtule commented:
@lowjoel I have fixed the bug and moved the logic to Course::Condition. Please take a look.

lowjoel commented:
Looks good, final question: is there a spec that tests the logic bug you fixed? Like when multiple conditional classes share the same condition class?

minhtule commented:
Sorry, didn't add test for that because it is a private method. To test the public method for that case, I would need to create to many database records. Is it possible to test private method in Rspec?

lowjoel commented:
This is questionable, but as practice here if the private method has sufficiently complex behaviour then it should be tested.

Use `send` to overcome the access visibility.

minhtule commented:
@lowjoel Just added the test [here](https://github.com/Coursemology/coursemology2/pull/699/files#diff-145cc44b149ae52e52634c82125218e2R42). Thank you for the `send` tip!

lowjoel commented on spec/models/course/condition_spec.rb:
```diff
@@ -3,4 +3,55 @@
+            to receive(:dependent_classes).and_return([Course::Achievement.name])
+          allow(Course::Condition::Assessment).
+            to receive(:dependent_classes).and_return([Course::Achievement.name])
+          actual_mapping = Course::Condition.send('dependent_class_to_condition_class_mapping')
```
by convention, this is a symbol, `:dependent_class_to_condition_class_mapping`

lowjoel commented:
Last thing, then it'll be merged.

minhtule commented:
Fixed!
